File: LatinReader/app/src/main/java/com/ericmschmidt/latinreader/datamodel/Dictionary.java
package com.ericmschmidt.latinreader.datamodel;

import android.util.Log;

import com.ericmschmidt.latinreader.utilities.DictionaryXMLHelper;
import com.ericmschmidt.latinreader.MyApplication;
import com.ericmschmidt.latinreader.utilities.ITextConverter;
import com.ericmschmidt.latinreader.utilities.ResourceHelper;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Random;

/**
 * Contains the data and methods for getting dictionary entries.
 */
public class Dictionary  {

    private ArrayList<String>  _entryHeaders;
    private WorkInfo dictionaryInfo;
    private ITextConverter _converter;

    /**
     * Creates a new instance of the dictionary class.
     */
    public Dictionary() {
        this.dictionaryInfo = MyApplication.getManifest().getDictionaryInfo();
        initEntries();
    }

    /**
     * Creates a new instance of the Dictionary class
     * and instantiates a text converter for non-Latin orthographies.
     * @param converter the ITextConverter to use
     */
    public Dictionary(ITextConverter converter) {
        this();
        this._converter = converter;
    }

    /**
     * Gets the number of entries in this dictionary.
     * @return integer
     */
    public int getEntryCount() {
        return this._entryHeaders.size();
    }

    /**
     * Checks whether a specific string is an entry in the dictionary.
     * @param searchEntry the entry to check for.
     * @return boolean
     */
    public boolean isInDictionary(String searchEntry) {
        return this._entryHeaders.contains(searchEntry);
    }

    /**
     * Gets a definition from the dictionary.
     * @param searchEntry the entry to search for.
     * @return String
     */
    public String getEntry(String searchEntry) {
        String definition = null;
        try {
            if (isInDictionary(searchEntry)) {
                InputStream stream = ResourceHelper.getResourceStream(this.dictionaryInfo.getLocation());
                definition = DictionaryXMLHelper.getEntry(stream, searchEntry, _converter);
            }
        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        }
        return definition;
    }

    /**
     * Gets a randomly selected from the dictionary
     * @return String randomly selected dictionary entry.
     */
    public String getRandomEntry() {
        Random random = new Random();
        int numberOfEntries = this.getEntryCount();

        int randomNumber = random.nextInt(numberOfEntries + 1);
        String randomEntryKey = this._entryHeaders.get(randomNumber);
        return getEntry(randomEntryKey);
    }

    // Gets the number of alphabet chapters in dictionary.
    private void initEntries() {
        try {
            Manifest manifest = MyApplication.getManifest();
            InputStream stream = ResourceHelper.getResourceStream(manifest.getDictionaryEntryResource());
            this._entryHeaders = DictionaryXMLHelper.getEntryHeaders(stream);

        } catch (Exception ex) {
            String errorMessage = ex.getMessage();
            MyApplication.logError(errorMessage);
        }
    }
}

File: LatinReader/core/src/main/java/com/ericmschmidt/classicsreader/datamodel/Dictionary.kt
package com.ericmschmidt.classicsreader.datamodel

import com.ericmschmidt.classicsreader.MyApplication.Factory.applicationInstance
import com.ericmschmidt.classicsreader.logError
import com.ericmschmidt.classicsreader.utilities.DictionaryXMLHelper
import com.ericmschmidt.classicsreader.utilities.ITextConverter
import com.ericmschmidt.classicsreader.utilities.ResourceHelper
import java.util.Random

/** Contains the data and methods for getting dictionary entries.
 * @author Eric Schmidt
 * @author <a href="https://telpirion.com">...</a>
 * @version 1.5
 * @since 1.1
 */
class Dictionary(val converter: ITextConverter? = null) {

    val manifest: Manifest = applicationInstance().manifest
    val dictionaryInfo: WorkInfo? = manifest.getDictionaryInfo()
    var entryHeaders: ArrayList<String> = ArrayList()

    init {
       initEntries()
    }

    /**
     * Gets the number of entries in this dictionary.
     */
    fun entryCount(): Int {
        return this.entryHeaders.size
    }

    /**
     * Checks whether a specific string is an entry in the dictionary.
     */
    fun isInDictionary(searchEntry: String?): Boolean {
        return this.entryHeaders.contains(searchEntry)
    }

    /**
     * Gets a definition from the dictionary.
     */
    fun getEntry(searchEntry: String?): String? {
        var definition: String? = ""

        if (!isInDictionary(searchEntry)) {
            return definition
        }

        ResourceHelper.getResourceStream(this.dictionaryInfo?.location as Int).use { stream ->
            definition = DictionaryXMLHelper.getEntry(stream, searchEntry, converter)
        }

        return definition
    }

    /**
     * Gets a randomly selected from the dictionary.
     */
    fun getRandomEntry(): String? {
        val random = Random()
        val numberOfEntries: Int = entryCount()

        val randomNumber = random.nextInt(numberOfEntries + 1)
        val randomEntryKey: String = this.entryHeaders[randomNumber]
        return getEntry(randomEntryKey)
    }

    // Gets the number of alphabet chapters in dictionary.
    private fun initEntries() {
        try {
            val stream = ResourceHelper.getResourceStream(manifest.getDictionaryEntryResource())
            this.entryHeaders = DictionaryXMLHelper.getEntryHeaders(stream)
        } catch (ex: java.lang.Exception) {
            val errorMessage = ex.message
            logError(errorMessage)
        }
    }
}